### ISSUE TITLE
Cypress crash: Error: Failed to parse or set cookie named

### ISSUE THREAD

**Issue by corneliutusnea** [cypress-io/cypress#1321]
- Operating System: Windows 10
- Cypress Version: 1.4.2
- Browser Version: Chrome 63?

### Is this a Feature or Bug?
Bug

### Current behavior:
Sometimes when I go to my website using cy.visit(...) if the site returns a redirect and a clear cookie (e.g. logout) the Cypress crashes and logs:
```
`{ [Error: Failed to parse or set cookie named "aspAuth".] name: undefined }
undefined
```

### How to reproduce:
Not sure, but these are the relevant headers my site returns when I navigate to it:

```
Cache-Control:no-cache, no-store, must-revalidate
Content-Length:0
Date:Fri, 16 Feb 2018 03:37:23 GMT
Expires:-1
Location:https://appcenter.intuit.com/connect/oauth2?...
p3p:CP="Company does not have a P3P policy"
Pragma:no-cache
Set-Cookie:Provision.ConnectionInfo=; expires=Thu, 01 Jan 1970 00:00:00 GMT; path=/; samesite=lax
Set-Cookie:XSRF-TOKEN=rzOLRz36OX_A5wQuhVPKkssxg7fCwZOa_iMi-Nx95-1RRU; path=/; secure; samesite=lax
Set-Cookie:XSRF-TOKEN=; expires=Thu, 01 Jan 1970 00:00:00 GMT; path=/; samesite=lax
Set-Cookie:aspAuth=; expires=Thu, 01 Jan 1970 00:00:00 GMT; path=/; samesite=lax
Strict-Transport-Security:max-age=31536000
X-Content-Type-Options:nosniff
X-Request-Id:0HLBKQPQAPOUV:00000003
X-Time:1
X-XSS-Protection:1
```

#### Test code:

```js
cy.visit('https://localdev/start/app');
```

**Comment by corneliutusnea** [comment]
So, I can reproduce this quite easily. Just try to delete/expire a cookie that was never set from the server and it will crash the complete Cypress app.
```
Set-Cookie: Name=; expires=Thu, 01 Jan 1970 00:00:00 GMT; path=/; samesite=lax
```

**Comment by bahmutov** [comment]
Thank you for the reproducible example

**Comment by lukemadera** [comment]
I have this issue as well. @corneliutusnea did you find a workaround? It's making cypress unusable.

**Comment by lukemadera** [comment]
Thanks for the prompt rely. @bahmutov I don't intend to give up that easily, but do need a fix - any progress on this?

**Comment by brian-mann** [comment]
We'll fix this in the next release.

**Comment by lukemadera** [comment]
Great, thanks Brian! Any ETA on when that will be?

**Comment by brian-mann** [comment]
Not sure yet - if the 3.0 is delayed any longer than this week we'll issue a patch release with these fixse

**Comment by lukemadera** [comment]
Awesome thanks - so it seems either way it should be in by the end of the week!

**Comment by lsplogi** [comment]
Is the patch released? I'm running into the same problem

**Comment by robertguss** [comment]
I too am having this same issue.

**Comment by darmalovan** [comment]
Having the same problem.

**Comment by brian-mann** [comment]
I looked into this issue and cannot reproduce. There is not enough information in the issue.

I tried recreating by setting cookies during a redirect which do not exist. This does not create the crash. To reproduce this we will need **a full reproducible example either by specifying the exact cypress commands, server responses (including all headers), redirects, etc**. 

Please try to minimize to the bare minimum. For instance, this does not cause a crash:

```js
// server side code
app.get("/redirect", (req, res) => {
  res
  .set({
    "Set-Cookie": "aspAuth=; expires=Thu, 01 Jan 1970 00:00:00 GMT; path=/; samesite=lax"
  })
  .redirect("http://localhost:3501/fixtures/dom.html")
})

// cypress code
cy.visit("/redirect")
```

**Comment by brian-mann** [comment]
I tried various combinations related to setting cookies and their values and none of them cause the crash.

**Comment by brian-mann** [comment]
@darmalovan please post the steps to reproduce... I cannot guarantee I will look at this issue again for awhile.

**Comment by brian-mann** [comment]
If nobody posts reproducible steps this issue will not be fixed. Really... anyone...?

**Comment by lukemadera** [comment]
For me it was trigger via logout and that call was not strictly necessary so I just removed it so found a work around for now. Additionally, now I'm getting a cookie too large / can't clear cookies issue that is failing the tests sooner so I can't even get anywhere, so Cypress has been down / blocked for me completely. See https://github.com/cypress-io/cypress/issues/1547
@brian-mann how do I manually set cookies / NOT have Cypress automatically set them for me?

**Comment by itaykotler-fundbox** [comment]
@lukemadera, @corneliutusnea, I had the same problem and figured a workaround:
`cy.window().then((win) => {
        win.location = "/logout";
  });`

**Comment by lujianlujian** [comment]
I am also having the same issue.
Cypress 2.1.0
Chrome 66
OSX 10.12.6

Just run this test and it crash:

```
describe('crash issue https://github.com/cypress-io/cypress/issues/1321', function() {
    it.only('go into create step2', function() {
      cy.visit('https://www.surveymonkey.com') //working fine
      cy.visit('https://www.surveymonkey.com/create/?sm=JyO9zQ7UD4cy1qYivEKDxbW5vSI_2BIVzBnqW7paZOcEU_3D') //crashing when trying to access a URL requiring login
    })
})
```
Would really like this to be fixed. Attached a debug log here: https://gist.github.com/lujianlujian/aa2f4aa22c194c55c776d6cc1421bb5c

**Comment by sillpa-polasani** [comment]
describe('My First Test', function() {
  it('Visits the Kitchen Sink', function() {
    cy.visit('https://www.paypal.com')
  })
})

{ [Error: Failed to parse or set cookie named "X-PP-SILOVER".] name: undefined }

I get this error - any idea how to solve this ?

**Comment by HeyDaniyar** [comment]
Cypress 2.1.0
Chrome 66
OSX 10.12.6

I am also having the same issue. The issue can be reproduced  exactly by the example of  @lujianlujian provides.

And seems it not only affect `cy.visit()`, `cy.request()` would also make cypress crashed or failed.

 After some local experiments, I found this issues happened for the request that its **response headers have "Set-Cookies"** attributes. And in some situations, the cypress would not crash but show the same error in the console.  The full log for this error is as bellow:

```
Error:     CypressError: cy.request() failed trying to load:

https://xxx/api/order/campaigns/xxx

We attempted to make an http request to this URL but the request failed without a response.

We received this error at the network level:

  > Failed to parse or set cookie named "xso".

-----------------------------------------------------------

The request we sent was:

Method: GET
URL: https://xxx/api/order/campaigns/xxx

-----------------------------------------------------------

Common situations why this would fail:
  - you don't have internet access
  - you forgot to run / boot your web server
  - your web server isn't accessible
  - you have weird network configuration settings on your computer

The stack trace for this error is:

undefined
```

**Comment by HeyDaniyar** [comment]
Cypress 2.1.0
Chrome 66
OSX 10.12.6

### How to reproduce

As @sillpa-polasani and @lujianlujian  describes, this crash can be reproduced by these cases:

```
describe('My First Test', function() {
  it('Visits the Kitchen Sink', function() {
    cy.visit('https://www.paypal.com')
  })
})

describe('My First Test', function() {
  it('Visits the Kitchen Sink', function() {
    cy.visit('https://www.paypal.com')
  })
})
```
### Workaround:

Based on some local debugging, I found this issue is mostly triggered by the unexpected promise result from  `setJarCookie` function. Therefore, the first workarond is add 
path: node_modules/cypress/dist/Cypress.app/Contents/Resources/app/packages/server/lib/request.js
```js
 setJarCookies: function(jar, automationFn) {
        var setCookie;
        setCookie = function(cookie) {
          var ex;
          cookie.name = cookie.key;
          if (cookie.name && cookie.name.startsWith("__cypress")) {
            return;
          }
          switch (false) {
            case cookie.maxAge == null:
              cookie.expiry = moment().unix() + cookie.maxAge;
              break;
            case !(ex = cookie.expires):
              cookie.expiry = moment(ex).unix();
          }
          return automationFn("set:cookie", cookie).then(function() {
            return Cookies.normalizeCookieProps(cookie);
          // when add this to catch error, the request function would work normally
          // but  it will still show the error in  cypress node console
          }).catch(function(err) {
            console.log('**** Get Error in Set Coookie *****', err);
          });
        };
        return Promise["try"](function() {
          var store;
          store = jar.toJSON();
          return Promise.each(store.cookies, setCookie);
        });
      },
```

And I suspect this unexpected error was related to cookie expires time, because after some hard code in automation set-cookie function, all things just work fine:

``` js
 setCookie: function(data, automate) {
        var cookie;
        if (isNamespaced(data)) {
          throw new Error("Sorry, you cannot set a Cypress namespaced cookie.");
        } else {
          cookie = normalizeCookieProps(data);
          cookie.url = extension.getCookieUrl(data);
          if (data.hostOnly) {
            cookie = _.omit(cookie, "domain");
          }
    
          /*  avoid setting some unexpected cooke in some situations */
          var currentTime = moment();
          var expires = data.expires;
          var expiry = data.expiry;
          if(( expiry && !expires) || (!expiry && expires)) {
            return;
          }

          if(expires &&  moment(expires) < currentTime ) {
            return;
          }
          return automate(cookie).then(normalizeCookieProps);
        }
      }


```

**Comment by brian-mann** [comment]
I have spent several hours tonight reproducing this. I appreciate you providing a reproducible example - although many of our users mentioned having this problem we could not easily reproduce on our own.

Problem 1:
I have yet to determine the **exact** root cause. Yes, cookies are failing to be set, but the actual reason why is much more complicated and is happening due to the way Cypress is handling visit resolution much earlier in the process.

Problem 2:
I'm also wondering why those failed promises aren't being correctly handled / caught. It appears that all Promises are attached correctly and therefore the exception should be handled and sent back to the runner correctly. This error should not be crashing Cypress.

**Comment by lukemadera** [comment]
Thank you so much for putting all that time and effort in Brian! Sorry this is such a tricky issue.

**Comment by tamtamchik** [comment]
@brian-mann did you see issue #1264? Might be a clue.

**Comment by brian-mann** [comment]
We're aware of what the problem is. It's not a quick fix but it's an important one that's high up on our radar.

We are going to release `3.1` first and then circle back to this. It requires shifting around a bunch of things to fix.

**Comment by debmalya** [comment]
I just landed in cypress.io and met the above error. 

If i run it through Electron 59 , did not get this issue. Only faced it with "Chrome 67".

**Comment by xlenz** [comment]
@debmalya nope, same issue in chrome 68

**Comment by hiredgun** [comment]
I've got the same problem :( Chrome 68

**Comment by Azeirah** [comment]
@HeyDaniyar I'd like to mention that I encountered the same problem, receiving "failed to parse or set ...". I used your two code snippets, but they failed due to the `moment` library missing.

I monkey-patched them to look like this:

```js
setCookie: function(data, automate) {
  var cookie;
  if (isNamespaced(data)) {
    throw new Error("Sorry, you cannot set a Cypress namespaced cookie.");
  } else {
    cookie = normalizeCookieProps(data);
    cookie.url = extension.getCookieUrl(data);
    if (data.hostOnly) {
      cookie = _.omit(cookie, "domain");
    }

    /* hard code start */
    var currentTime = +new Date();
    var expires = data.expires;
    var expiry = data.expiry;
    if(( expiry && !expires) || (!expiry && expires)) {
      return;
    }

    if(expires && +new Date(expires) < currentTime ) {
      return;
    }
    /* hard code ends */
    return automate(cookie).then(normalizeCookieProps);
  }
},
```



```js
setJarCookies: function(jar, automationFn) {
  var setCookie;
  setCookie = function(cookie) {
    var ex;
    cookie.name = cookie.key;
    if (cookie.name && cookie.name.startsWith("__cypress")) {
      return;
    }
    switch (false) {
      case cookie.maxAge == null:
        cookie.expiry = moment().unix() + cookie.maxAge;
        break;
      case !(ex = cookie.expires):
        cookie.expiry = moment(ex).unix();
    }
    return automationFn("set:cookie", cookie).then(function() {
      return Cookies.normalizeCookieProps(cookie);
     // when add catch funtion to the promise,
     // the request function would return result normally.
     // But  it will still show the error in  cypress node console
    }).catch(function(err) {
      console.log('**** Get Error in Set Coookie *****', err);
    });
  };
  return Promise["try"](function() {
    var store;
    store = jar.toJSON();
    return Promise.each(store.cookies, setCookie);
  });
},
```

**Comment by kentcdodds** [comment]
For those landing here after 3.1.0 was released, the location of the Cypress executable (Cypress.app) was changed and you can find the new location by running `npx cypress verify` and it'll print the path :)

For me it's here: `/Users/kdodds/Library/Caches/Cypress/3.1.0/Cypress.app`

**Comment by cannibalcow** [comment]
Does anyone have a workaround for this bug. It's a showstopper for my team right now.  We are on the brink of abandoning cypress for this.

**Comment by xlenz** [comment]
@cannibalcow you'd better abandon if you can, regular webdriver is much much better. It takes time on the beginning to setup things, but then it works way better and faster.

Cypress is good only for something really small. It is good on the first step, but then it is much more painful to work with. You'll have lot's of showstoppers

And it support headless chrome.

**Comment by HeyDaniyar** [comment]
@cannibalcow Have you ever tried my workaround?  At least now it works for me.

**Comment by drod3763** [comment]
I'm able to reproduce this as well. It just so happens it fails to parse on a Cookie with a expiration date in the past.

**Comment by vincentvb** [comment]
I can also confirm that this issue is stopping my team from getting our cypress test suite off the ground. I know a fix is in the works, but is there an ETA by chance?

**Comment by vinhlh** [comment]
I have this issue when having consecutive visit commands,
```js
cy.visit('/pageA')
  .doSomething()
  .visit('/anotherPage')
```
 and it's also a bad practice when implenting e2e tests, then I just change it to clicking UI items.

**Comment by brian-mann** [comment]
We're prioritizing this issue in this sprint. It's long overdue to be fixed.

**Comment by brian-mann** [comment]
We have made a lot of progress on this issue, and may have a fix out in the next day or two. We had to rewrite the visit resolving layer of Cypress, which will help more things than just this bug as well.

**Comment by paglajewel** [comment]
Hi @brian-mann, we are doing a POC on using Cypress for our apps. We are also running into this issue. What is the eta on releasing this fix? You mentioned 3 days ago "may have a fix out in the next day or two". Its a blocker for us.

**Comment by drewbrend** [comment]
I also ran into this issue, but encountered it when sending a POST cy.request to authenticate with our SSO provider - okta.  Looks like someone else hit the same blocker as well:  https://github.com/cypress-io/cypress/issues/1489#issuecomment-418871535

**Comment by paglajewel** [comment]
Hey @drewbrend, you can try password grant type in okta if you are doing via cy.request. https://developer.okta.com/blog/2018/06/29/what-is-the-oauth2-password-grant

**Comment by jennifer-shehane** [comment]
The code for this is done, but this has yet to be released. We'll update this issue and reference the [changelog](https://on.cypress.io/changelog) when it's released.

**Comment by Ivan-Cristian** [comment]
> @cannibalcow Have you ever tried my workaround? At least now it works for me.

Would you mind sharing what your workaround is?

**Comment by jm2242** [comment]
Interestingly, We are now getting this problem (or similar) with the upgrade to 3.3.1. Everything works in 3.3.0 for us. We are getting this error in a few scenarios. One, whenever we try to log in:


```
~/Quorum/quorum-site on  cypress-docs! ⌚ 11:33:02
$ npm run cypress                                                                                                                                                                              ‹ruby-2.4.1›

> quorum-site@2.8.3 cypress /Users/jonathanmares/Quorum/quorum-site
> cypress open

GET /__/ 200 28.124 ms - -
GET /__cypress/runner/cypress_runner.css 200 52.233 ms - -
GET /__cypress/runner/cypress_runner.js 200 383.054 ms - -
GET /__cypress/runner/fonts/fontawesome-webfont.woff2?v=4.6.3 200 1.634 ms - 71896
GET /__cypress/iframes/integration/main/login_spec.js 200 14.349 ms - 733
GET /__cypress/tests?p=cypress/integration/main/login_spec.js-216 200 682.720 ms - -
GET /__cypress/tests?p=cypress/support/index.js-177 200 840.621 ms - -
{ [Error: Failed to parse or set cookie named "current_version".] name: undefined }
undefined
(quorum)
```

This is our `index.js` file:
```
// cypress/support/index.js
// uncomment below to have Cypress log when Cookies change
Cypress.Cookies.debug(true)

// Make it possible to skip auth across tests
Cypress.Cookies.defaults({
    whitelist: ["qsesid", "csrftoken"],
})
```
(interestingly, with our without the defaults, we still get the crash. 
Our login configuration is the following:
```
Cypress.Commands.add('loginByCSRF', ({csrfToken, username, password}) => {
    cy.request({
        method: 'POST',
        url: '/login/',
        failOnStatusCode: false, // dont fail so we can make assertions
        form: true, // we are submitting a regular form body
        body: {
            username,
            password,
        },
        headers: {
            "X-CSRFTOKEN": csrfToken,
            // django spells this referer - very odd
            // this is required for Django to let this endpoint authenticate
            "referer": Cypress.config().baseUrl
        }
    })
})

/**
 * This function ("login") logs in to Quorum programmatically
 * @name login
 * @function
 * @param {String} username - a username to login with
 * @param {String} password - a password to log in with
 * @param {Boolean} useCurrentSessionIfPossible - if qsesid and csrftoken are defined, don't log in again
 * @param {Boolean} shouldVisitPath - Provide a way to skip visiting path if desired
 *
 * You might want to pass shouldVisitPath=false if you are refreshing the page in your beforeEach() block,
 *  in which case the initial visit is wasteful
 */
Cypress.Commands.add(
    "login",
    ({
        username = Cypress.config("TESTING_USER"),
        password = Cypress.config("TESTING_PASSWORD"),
        path = "/home/",
        shouldVisitPath = true,
        useCurrentSessionIfPossible = false,
    } = {}) => {
    if (useCurrentSessionIfPossible && cy.getCookie("qsesid").value && cy.getCookie("csrftoken").value) {
        console.log("Already logged in, not logging in again")
        return
    }
    // make a request to get a csrftoken
    cy.request("/login")
        .its("headers")
        .then((headers) => {
            const csrfToken = headers["set-cookie"][0].split(";")[0].replace("csrftoken=", "")
            cy.loginByCSRF({ csrfToken, username, password }).then(() => {
                shouldVisitPath && path && cy.visit(path)
            })
        })
})
```

here's the beginning of a test file that fails:
```
describe("Home page tests", function () {
    before(function() {

        // make sure we're logged in. Let beforeEach() handle visiting the home page
        cy.login({
            shouldVisitPath: false,
            useCurrentSessionIfPossible: true // if auth cookies are available, don't bother re-logging in
        })
    })
    beforeEach(function () {

        // set up server to listen for requests
        cy.server()

        // alias requests to newperson so we can wait until they resolve
        cy.route("GET", "/api/newperson/**").as("getNewPerson")

        // before each test, go to the home page
        cy.visit("/")
    })
```


And also we don't log in, but this time it doesn't crash, it just gives an error in the GUI:
![image](https://user-images.githubusercontent.com/8572502/48276966-55394f00-e417-11e8-9a30-231e4892f545.png)

**Comment by jm2242** [comment]
@jennifer-shehane doesn't like I have permission to re-open the issue, should I make a separate issue for this?

**Comment by jennifer-shehane** [comment]
We're aware of a regression with cookies in 3.1.1 here: https://github.com/cypress-io/cypress/issues/2724 not sure if this is related.

**Comment by jm2242** [comment]
We're still getting this error in all versions after `3.1.0`, including latest `3.1.3` release

**Comment by ChaseWNorton** [comment]
Do we have an update on this bug? Nov 1 seemed like it was close to release - we are still experiencing this in 3.1.3

**Comment by jennifer-shehane** [comment]
The fix for this was released in version `3.1.1`. If you are experiencing this issue, we will need a reproducible example in version 3.1.3 provided - or a new issue opened if it does not reflect the exact same error as this issue above.

**Comment by jm2242** [comment]
@jennifer-shehane post above describes our setup. Is there something I should add?

**Comment by jennifer-shehane** [comment]
@jm2242 We need an example that we can run locally to see the exact crashing behavior. Copying and pasting the code provided in your example will not work in our environment - since it's dependent on your localhost running and you haven't provided the application code.

**Comment by randing89** [comment]
Still seeing this issue in 3.1.5. 

`Error: Failed to parse or set cookie named "JSESSIONID".`

This is the cookie in the header.
```
set-cookie: JSESSIONID=4C30AC94E795759EC164FF34D2FD5AB6; Path=/wiki; Secure; HttpOnly
```

**Comment by sweoggy** [comment]
Can confirm @randing89's finding that Secure keyword breaks Cypress 3.1.5 (worked fine in 3.1.0).

**Comment by mutoe** [comment]
> Can confirm @randing89's finding that Secure keyword breaks Cypress 3.1.5 (worked fine in 3.1.0).

@randing89 @sweoggy Thanks, I downgraded to 3.1.0 and it works fine. 😄

**Comment by jennifer-shehane** [comment]
@randing89 @sweoggy @mutoe Could one of you open an issue detailing the new problem in 3.1.5 please?

**Comment by ArtanisAce** [comment]
Also happening to me. It doesn't happen locally but happens when pointing to our DEV servers (also http, so don't really get it). 3.1.0 only throws error when doing `cy.logout()` (probably some issue destroying cookies??), and 3.1.5 throws this cookies error almost always. So, with 3.1.0 and this logout replacement by @itaykotler-fundbox works a bit better, but still I'm seeing that I get the error on some cases. Hope there's a fix for this soon :/

**Comment by Bonfipa** [comment]
Hello, sorry, my English is not fluent so I'm using google translator. After a few days suffering, doing everything I saw in the forums and could not solve the problem.

How I solved: In my tests I call a login function before each case, and there was the problem, I created a suite of tetes for login screen, and the others I do the direct access. From what I understand, since I do not log off at the end of the tests, then it raises the problem. I'm a beginner in automated testing, but that's how I can solve it.
Cypress version: 3.4.0
